### PR TITLE
fixes #1697 shard travis builds for python2

### LINKED ISSUE #1697: shard travis builds for python 2
# Issue Type

- Bug report

# Desired Behavior

Travis builds for python 2.7 should get sharded as the builds for python 3

# Actual Behaviour

Builds are not sharded, and sometimes it causes time limit error.

```
The job exceeded the maximum time limit for jobs, and has been terminated.
```

Example: https://travis-ci.com/ansible/molecule/jobs/172385426

### PR DESCRIPTION
**Pull request by themr0c**
Signed-off-by: Fabrice Flore-Thebault <themr0c@users.noreply.github.com>

fixes #1697 shard travis builds for python2

#### PR Type

- Bugfix Pull Request

### CODE REVIEW

**Comment by themr0c** [comment]
NB: if it appears to be impossible, there is already some ticket to propose some more deep enhancement #1575

File: .travis.yml [hunk 1]
```diff
@@ -48,6 +48,16 @@ language: python
 - &env-functional-shard_3
   <<: *env-functional
   PYTEST_ADDOPTS: "'-m shard_3_of_3'"
+- &py27-env-functional-shard_1
```
Comment (webknjaz) [inline .travis.yml]: This set of YAML anchors is not needed: those above hold exact copies of the same values as below.
Also, you call them "py27" while there's literally nothing pointing to python 2.7, they are as generic as ones without such prefix.

File: .travis.yml [hunk 2]
```diff
@@ -72,164 +82,229 @@ jobs:
     env:
       ANSIBLE_VERSION: "27"
       TESTS_TYPE: unit
+    name: unit tests, Ansible 2.7, Python 3.7
```
Comment (webknjaz) [inline .travis.yml]: I'd prefer having UI changes in a separate PR. Having a bunch of mixed patches in one PR makes it non-atomic and hard to review as it's fairly complicated to find bits which are worth checking and easy to miss important changes.
So please let's avoid having such white noise in patches. Also, this change is clearly not "fixes #1697 shard travis builds for python2"

File: .travis.yml [hunk 3]
```diff
@@ -72,164 +82,229 @@ jobs:
...
   - <<: *py-36
     <<: *if-cron-or-manual-run-or-tagged
     env:
       <<: *env-functional-shard_3
       ANSIBLE_VERSION: "25"
+    name: functional tests shard 3/3, Ansible 2.5, Python 3.6
+
+  - <<: *py-27
+    env:
+      <<: *py27-env-functional-shard_1
```
Comment (webknjaz) [inline .travis.yml]:
```suggestion
      <<: *env-functional-shard_1
```

File: .travis.yml [hunk 4]
```diff
@@ -72,164 +82,229 @@ jobs:
...
+      ANSIBLE_VERSION: "27"
+    name: functional tests shard 1/3, Ansible 2.7, Python 2.7
 
   - <<: *py-27
     env:
+      <<: *py27-env-functional-shard_2
```
Comment (webknjaz) [inline .travis.yml]:
```suggestion
      <<: *env-functional-shard_2
```

File: .travis.yml [hunk 5]
```diff
@@ -72,164 +82,229 @@ jobs:
...
       ANSIBLE_VERSION: "27"
-      TESTS_TYPE: functional
+    name: functional tests shard 2/3, Ansible 2.7, Python 2.7
 
   - <<: *py-27
     env:
+      <<: *py27-env-functional-shard_3
```
Comment (webknjaz) [inline .travis.yml]:
```suggestion
      <<: *env-functional-shard_3
```

File: .travis.yml [hunk 6]
```diff
@@ -72,164 +82,229 @@ jobs:
...
+      ANSIBLE_VERSION: "27"
+    name: functional tests shard 3/3, Ansible 2.7, Python 2.7
+
+  - <<: *py-27
+    env:
+      <<: *py27-env-functional-shard_1
       ANSIBLE_VERSION: "26"
-      TESTS_TYPE: functional
```
Comment (webknjaz) [inline .travis.yml]: Removal of such env var WILL BREAK construction of `TOXENV` var. Don't do this.

File: .travis.yml (hunk 2 above)
Comment (webknjaz) [reply .travis.yml]: P.S. I've extracted names into the master

File: .travis.yml (hunk 1 above)
Comment (themr0c) [reply .travis.yml]: I am really happy to find someone pickier than me \o/.
PYTEST_ADDOPTS: "'-m shard_3_of_3'", with single quotes, is functional wit python3, so i didn't change it. But it didn't work with python2
The attempt to remove the single quotes failed miserably PYTEST_ADDOPTS: "-m shard_3_of_3"
The current attempt to without the single quotes but with an = in place of the space appears to be functional PYTEST_ADDOPTS: "-m=shard_3_of_3".
Indeed, now, good question: is it better to keep an exception for py2, or is it better to, at least, try the "-m=shard_3_of_3" for all versions pf python?

File: .travis.yml (hunk 6 above)
Comment (themr0c) [reply .travis.yml]: Isn'tit done in py27-env-functional-shard_1 ?

**Comment by themr0c** [comment]
@webknjaz Thank you for your comments, much appreciated. I replied to some, and I will adapt after you confirm the final direction to take.

File: .travis.yml (hunk 1 above)
Comment (webknjaz) [reply .travis.yml]: I'm fine with both, I just wanted to reduce duplication which leads to even more diverse semi-copypasted approaches in future

File: .travis.yml (hunk 2 above)
Comment (webknjaz) [reply .travis.yml]: Oh, the conflict is not yet resolved

File: .travis.yml (hunk 6 above)
Comment (webknjaz) [reply .travis.yml]: Oh, right.. I overlooked it

File: .travis.yml (hunk 3 above)
Comment (themr0c) [reply .travis.yml]: done

File: .travis.yml (hunk 4 above)
Comment (themr0c) [reply .travis.yml]: done

File: .travis.yml [hunk 7]
```diff
@@ -244,22 +247,57 @@ jobs:
...
+      ANSIBLE_VERSION: "26"
+    name: functional tests shard 2/3, Ansible 2.6, Python 2.7
+
+  - <<: *py-27
+    env:
+      <<: *env-functional-shard_3
+      ANSIBLE_VERSION: "26"
+    name: functional tests shard 3/3, Ansible 2.6, Python 2.7
+
+ - <<: *py-27
```
Comment (webknjaz) [inline .travis.yml]: Add a space in front of this

Comment (themr0c) [reply .travis.yml]: i fixed this and all yamllint errors

**Comment by themr0c** [comment]
Good news the errors on the last build https://travis-ci.com/ansible/molecule/builds/98812943 are errors that are already in master, unrelated to this PR.

File: .travis.yml [hunk 8]
```diff
@@ -8,369 +9,407 @@ cache:
   bundler: true
   pip: true
   directories:
-  - $HOME/.cache/pre-commit
-  - $HOME/.pre-commit
-  - $HOME/.rvm
-  - $HOME/virtualenv/python$(python -c 'import platform; print(platform.python_version())')
-  - $HOME/Library/Caches/Homebrew
+    - $HOME/.cache/pre-commit
```
Comment (webknjaz) [inline .travis.yml]: I like previous style more, I used it on purpose. It helps to maintain shorter lines even for nested struts

Comment (themr0c) [reply .travis.yml]: This is not a matter of opinion, the changes were requested by yamllint. I can rollback no problem.

Comment (gundalow) [reply .travis.yml]: Lets go with yamllint

**Comment by gundalow** [comment]
This improves on what we have, so merging
@themr0c @lwm @webknjaz Thank you all for your work on this

**Comment by webknjaz** [comment]
:+1:

**Pull request merged**